File: frappe_mpsa_payments/www/mpesa/stkpush.js
let checking = false;
let interval = null;
window.current_converted_amount = "{{ mpesa_request.amount or '0.00' }}";

function getRequestRedirect() {
	const p = new URLSearchParams(window.location.search);
	return p.get("redirect_to");
}

function getRequestId() {
	const p = new URLSearchParams(window.location.search);
	return p.get("id");
}

function showOverlay(text = "Processing...") {
	const o = document.getElementById("overlay");
	if (o) {
		o.querySelector(".mp-overlay-text").innerText = text;
		o.style.display = "flex";
	}
}

function hideOverlay() {
	const o = document.getElementById("overlay");
	if (o) o.style.display = "none";
}

frappe.ready(() => {
	["phone_number", "base_amount", "reference_id"].forEach((id) => {
		const el = document.getElementById(id);
		if (el && !el.disabled) {
			el.addEventListener("input", updateURL);
		}
	});

	const id = getRequestId();
	if (id) {
		beginChecking();
	} else {
		update_total_display();
	}
});

function updateURL() {
	const p = new URLSearchParams(window.location.search);
	const fields = ["phone_number", "base_amount", "reference_id"];

	fields.forEach((field) => {
		const el = document.getElementById(field);
		if (el && el.value) p.set(field, el.value);
		else p.delete(field);
	});

	history.replaceState({}, "", location.pathname + "?" + p.toString());
}

async function update_total_display() {
	const baseAmount = document.getElementById("base_amount")?.value;
	const currencyLabel = document.getElementById("currency_label");
	const currency = currencyLabel ? currencyLabel.innerText.trim() : "KES";
	const gatewayInput = document.getElementById("payment_gateway");
	const gatewayValue = gatewayInput ? gatewayInput.value : "";
	const displayElement = document.getElementById("total_display");

	if (!baseAmount || parseFloat(baseAmount) <= 0) {
		if (displayElement) displayElement.innerText = "Ksh 0.00";
		window.current_converted_amount = "0.00";
		return;
	}

	if (currency === "KES") {
		const formatted = parseFloat(baseAmount).toFixed(2);
		if (displayElement) displayElement.innerText = `Ksh ${formatted}`;
		window.current_converted_amount = formatted;
		return;
	}

	try {
		const r = await frappe.call({
			method: "frappe_mpsa_payments.utils.utils.convert_amount_to_kes",
			args: {
				amount: baseAmount,
				currency: currency,
				settings: gatewayValue.replace("Mpesa-", ""),
			},
		});
		if (r.message && displayElement) {
			const converted = parseFloat(r.message).toFixed(2);
			displayElement.innerText = `Ksh ${converted}`;
			window.current_converted_amount = converted;
		}
	} catch (e) {
		console.error("Currency conversion failed", e);
	}
}

async function submit_new_request() {
	const btn = document.getElementById("create-btn");

	const data = {
		phone_number: document.getElementById("phone_number")?.value,
		payment_gateway: document.getElementById("payment_gateway")?.value,
		reference_type: document.getElementById("reference_type")?.value,
		reference_id: document.getElementById("reference_id")?.value,
		base_amount: document.getElementById("base_amount")?.value,
		currency: "KES",
		amount: window.current_converted_amount,
		title: document.getElementById("title")?.innerText || "",
		description: document.getElementById("description")?.innerText || "",
	};

	if (!data.phone_number || !data.base_amount || !data.reference_id) {
		return frappe.msgprint(__("Please fill in all required fields."));
	}

	showOverlay("Sending STK Push...");
	if (btn) {
		btn.style.pointerEvents = "none";
		btn.innerHTML = "<span>Processing...</span>";
	}

	try {
		const res = await frappe.call({
			method: "frappe_mpsa_payments.frappe_mpsa_payments.doctype.mpesa_express_request.mpesa_express_request.create_new_request",
			args: data,
		});
		const redirect_to = getRequestRedirect();
		location.href =
			location.pathname +
			"?id=" +
			res.message +
			(redirect_to ? "&redirect_to=" + encodeURIComponent(redirect_to) : "");
	} catch (e) {
		hideOverlay();
		if (btn) {
			btn.style.pointerEvents = "all";
			btn.innerHTML = `<span>Pay Now</span><svg width="20" height="20" fill="none" stroke="currentColor" stroke-width="2" viewBox="0 0 24 24"><path d="M5 12h14M12 5l7 7-7 7"></path></svg>`;
		}
	}
}

function beginChecking() {
	if (checking) return;
	checking = true;
	showOverlay("Checking payment status...");
	checkStatus();
	interval = setInterval(checkStatus, 5000);
}

async function checkStatus() {
	if (!checking) return;
	const id = getRequestId();
	if (!id) return;

	const r = await frappe.call({
		method: "frappe_mpsa_payments.frappe_mpsa_payments.doctype.mpesa_express_request.mpesa_express_request.get_request_status",
		args: { name: id },
	});

	const status = r.message?.status;
	if (!status) return;

	const badge = document.getElementById("status-badge");
	if (badge) badge.innerText = status;

	if (status === "Completed") {
		stopChecking();
		const redirect_to = getRequestRedirect();
		if (redirect_to) window.location.href = redirect_to;
		else hideOverlay();
	}

	if (status === "Failed") {
		stopChecking();
		hideOverlay();
		const retryBtn = document.getElementById("retry-btn");
		if (retryBtn) retryBtn.style.display = "block";
	}
}

function stopChecking() {
	checking = false;
	clearInterval(interval);
}

function cancel_check() {
	stopChecking();
	hideOverlay();
}

async function retry_stk() {
	const id = getRequestId();
	if (!id) return;
	showOverlay("Retrying payment...");
	await frappe.call({
		method: "frappe_mpsa_payments.frappe_mpsa_payments.doctype.mpesa_express_request.mpesa_express_request.retry_stkpush",
		args: { name: id },
	});
	beginChecking();
}

File: frappe_mpsa_payments/www/mpesa/stkpush.py
import frappe
from frappe import _
from frappe_mpsa_payments.utils.utils import convert_amount_to_kes


def get_context(context):
    q = frappe.local.form_dict
    context.redirect_to = q.get("redirect_to")
    context.mpesa_gateways = frappe.get_all(
        "Payment Gateway",
        filters={"gateway_settings": ["is", "set"]},
        fields=["name"],
        ignore_permissions=True,
    )

    currency = q.get("currency") or "KES"
    context.currency = currency
    gateway = q.get("payment_gateway")
    setting_name = gateway.replace("Mpesa-", "") if gateway else None
    settings = frappe.get_doc("Mpesa Settings", setting_name) if setting_name else None

    context.title = settings.payment_page_title if settings and settings.payment_page_title else q.get("title")
    setting_desc = settings.payment_page_description if settings else None
    query_desc = q.get("description")
    context.description = f"{setting_desc} - {query_desc}" if setting_desc and query_desc else (setting_desc or query_desc)

    context.allow_amount_editing = settings.allow_amount_editing if settings else False
    context.allow_payment_reference_editing = settings.allow_payment_reference_editing if settings else False
    

    if q.get("id"):
        context.is_new = False
        load_existing(context, q.get("id"))
        frappe.local.form_dict = {"id": q.get("id"), "redirect_to": context.redirect_to}
    else:
        context.is_new = True
        base_amount = clean_null(q.get("base_amount") or q.get("amount"))
        actual_amount = base_amount

        if currency != "KES" and base_amount:
            try:
                actual_amount = convert_amount_to_kes(
                    amount=float(base_amount), currency=currency, settings= setting_name
                )
            except Exception as e:
                actual_amount = base_amount

        context.mpesa_request = {
            "phone_number": clean_null(q.get("phone_number")),
            "payment_gateway": gateway,
            "reference_type": clean_null(q.get("reference_type")),
            "reference_id": clean_null(q.get("reference_id")),
            "base_amount": base_amount,
            "currency": currency,
            "amount": actual_amount,
            "title": context.title,
            "description": context.description,
        }


def clean_null(v):
    if v in (None, "null", "", "undefined"):
        return ""
    return v


def load_existing(context, id):
    try:
        doc = frappe.get_doc("Mpesa Express Request", id)
        context.mpesa_request = {
            "phone_number": doc.phone_number,
            "payment_gateway": doc.payment_gateway,
            "reference_type": doc.reference_doctype,
            "reference_id": doc.reference_name,
            "base_amount": getattr(doc, "base_amount", doc.amount),
            "amount": doc.amount,
            "currency": getattr(doc, "currency", "KES"),
            "checkout_request_id": doc.checkout_request_id,
            "status": doc.status,
            "response_description": doc.response_description,
            "title": doc.transaction_title,
            "description": doc.transaction_description,
        }
        
        if doc.status == "Completed":
            context.success = _("Payment completed successfully")
        elif doc.status == "Failed":
            context.error = _("Payment failed: ") + (
                doc.response_description or "Unknown Error"
            )
    except frappe.DoesNotExistError:
        context.error = _("Request not found")
        context.is_new = True
        context.mpesa_request = {}
